pkg/snapshotter/snapshotter_csi.go

```diff
@@ -0,0 +1,464 @@
+	return nil
+}
+
+func (c *csiDriver) SnapshotStatus(name, namespace string) (SnapshotInfo, error) {
+	var snapshotInfo SnapshotInfo
+	snapshot, err := c.snapshotClient.SnapshotV1beta1().VolumeSnapshots(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+	if err != nil {
+		snapshotInfo.Status = StatusFailed
+		snapshotInfo.Reason = fmt.Sprintf("snapshot %s lost during backup: %v", name, err)
+		return snapshotInfo, err
+
+	}
+	snapshotInfo.SnapshotRequest = snapshot
+	var snapshotClassName string
+	if snapshot.Spec.VolumeSnapshotClassName != nil {
+		snapshotClassName = *snapshot.Spec.VolumeSnapshotClassName
+	}
+	snapshotClass, err := c.snapshotClient.SnapshotV1beta1().VolumeSnapshotClasses().Get(context.TODO(), snapshotClassName, metav1.GetOptions{})
+	if err != nil {
+		snapshotInfo.Status = StatusFailed
+		snapshotInfo.Reason = fmt.Sprintf("snapshot class %s lost during backup: %v", snapshotClassName, err)
+		return snapshotInfo, err
+	}
+
+	volumeSnapshotReady := c.snapshotReady(snapshot)
+	var volumeSnapshotContentReady bool
+	var snapshotContent *kSnapshotv1beta1.VolumeSnapshotContent
+	var contentName string
+	if volumeSnapshotReady && snapshot.Status.BoundVolumeSnapshotContentName != nil {
+		snapshotContentName := *snapshot.Status.BoundVolumeSnapshotContentName
+		snapshotContent, err = c.snapshotClient.SnapshotV1beta1().VolumeSnapshotContents().Get(context.TODO(), snapshotContentName, metav1.GetOptions{})
+		if err != nil {
+			snapshotInfo.Status = StatusFailed
+			snapshotInfo.Reason = fmt.Sprintf("snapshot content %s lost during backup: %v", snapshotClassName, err)
+			return snapshotInfo, err
+		}
+		snapshotInfo.Content = snapshotContent
+		snapshotInfo.Class = snapshotClass
+		volumeSnapshotContentReady = c.snapshotContentReady(snapshotContent)
+		contentName = snapshotContent.Name
+	}
+
+	// Evaluate current status of the backup for this PVC. Get all metadata and decide if finished.
+	var vsError string
+	if snapshot.Status != nil && snapshot.Status.Error != nil && snapshot.Status.Error.Message != nil {
+		vsError = *snapshot.Status.Error.Message
+	}
+	var pvcName string
+	if snapshot.Spec.Source.PersistentVolumeClaimName != nil {
+		pvcName = *snapshot.Spec.Source.PersistentVolumeClaimName
+	}
+	size := getSnapshotSize(snapshot)
+	if size == 0 && len(pvcName) != 0 {
+		// if restoreSize is empty, report PVC size
+		pvc, err := core.Instance().GetPersistentVolumeClaim(
+			pvcName,
+			namespace,
+		)
+		if err == nil {
+			size = getPVCSize(pvc)
+		}
+	}
+	var vscError string
+	if contentName != "" {
+		vscError = c.getSnapshotContentError(contentName)
+	}
+	switch {
+	case volumeSnapshotReady && volumeSnapshotContentReady:
+		snapshotInfo.Status = StatusReady
+		snapshotInfo.Reason = "Snapshot successful for volume"
+		snapshotInfo.Size = uint64(size)
+
+	case time.Now().After(snapshot.CreationTimestamp.Add(snapshotTimeout)):
```

Comment:
should we make `snapshotTimeout`  configurable by reading it as env variable, by defaults its 5 mins